File: algorithms/java/src/test/java/leetcode/LC080RemoveDuplicatesFromSortedArrayII.java
package leetcode;

import org.junit.jupiter.api.Test;


import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/
 * #medium
 */
public final class LC080RemoveDuplicatesFromSortedArrayII {
    public final class Solution {
        public int removeDuplicates(final int[] nums) {
            int i = 0;
            for (int n : nums)
                if (i < 2 || n > nums[i - 2])
                    nums[i++] = n;
            return i;
        }
    }

    @Test
    public void test_example() throws Exception {
        int[] nums = {1, 1, 1, 2, 2, 3};
        int[] expected = {1, 1, 2, 2, 3};
        assertEquals(5, new Solution().removeDuplicates(nums));
        assertArrayEquals(expected, Arrays.copyOf(nums, expected.length));
    }

    @Test
    public void test_1_1_1_1_3_3() throws Exception {
        int[] nums = {1, 1, 1, 1, 3, 3};
        int[] expected = {1, 1, 3, 3};
        assertEquals(4, new Solution().removeDuplicates(nums));
        assertArrayEquals(expected, Arrays.copyOf(nums, expected.length));
    }

    @Test
    public void test_1_1() throws Exception {
        int[] nums = {1, 1};
        int[] expected = {1, 1};
        assertEquals(2, new Solution().removeDuplicates(nums));
        assertArrayEquals(expected, Arrays.copyOf(nums, expected.length));
    }


    @Test
    public void test_1_2_2() throws Exception {
        int[] nums = {1, 2, 2};
        int[] expected = {1, 2, 2};
        assertEquals(3, new Solution().removeDuplicates(nums));
        assertArrayEquals(expected, Arrays.copyOf(nums, expected.length));
    }
}

File: algorithms/code/leetcode/lc080_remove_duplicates_from_sorted_array_ii/lc080_remove_duplicates_from_sorted_array_ii.rs
// https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/
// #medium

pub struct Solution;

impl Solution {
    #[allow(clippy::ptr_arg)]
    pub fn remove_duplicates(nums: &mut Vec<i32>) -> i32 {
        let mut i = 0;
        for num_index in 0..nums.len() {
            let n = nums[num_index];
            if i < 2 || n > nums[i - 2] {
                nums[i] = n;
                i += 1;
            }
        }
        i as i32
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let mut nums = vec![1, 1, 1, 2, 2, 3];
        let expected = vec![1, 1, 2, 2, 3];
        assert_eq!(Solution::remove_duplicates(&mut nums), 5);
        assert_eq!(nums[..expected.len()], expected[..]);
    }

    #[test]
    fn test_111133() {
        let mut nums = vec![1, 1, 1, 1, 3, 3];
        let expected = vec![1, 1, 3, 3];
        assert_eq!(Solution::remove_duplicates(&mut nums), 4);
        assert_eq!(nums[..expected.len()], expected[..]);
    }

    #[test]
    fn test_11() {
        let mut nums = vec![1, 1];
        let expected = vec![1, 1];
        assert_eq!(Solution::remove_duplicates(&mut nums), 2);
        assert_eq!(nums[..expected.len()], expected[..]);
    }

    #[test]
    fn test_122() {
        let mut nums = vec![1, 2, 2];
        let expected = vec![1, 2, 2];
        assert_eq!(Solution::remove_duplicates(&mut nums), 3);
        assert_eq!(nums[..expected.len()], expected[..]);
    }
}
